### PR TITLE
Add custom validation function testcases

### PR DESCRIPTION
**Pull request by xdarklight**
The new test ensures that promises can be returned by custom model
validation functions (there is already a test which does the same tests
for field-based custom validation functions).

### CODE REVIEW

**Comment by xdarklight** [comment]
There is a failing test which is completely unrelated to this PR:
```
  1) [POSTGRES-NATIVE] associations scope N:M on the through model should create, find and include associations with scope values:
```

**Comment by sushantdhiman** [comment]
You should move this test to `unit` tests, and modify it using `sinon` so it doesn't make any database call.

**Comment by xdarklight** [comment]
@sushantdhiman I added unit-tests for custom validation functions in general.

My initial commit adds an integration-test which runs a query from a custom validation function.
So I have two questions:
1. is this use-case not worth adding an integration-test?
2. are there any examples which test this promise behavior?
(in my case the custom validation function returns a promise to a function which then marks validation as failed - it's not the custom validation function which is failing directly)

**Comment by mickhansen** [comment]
@xdarklight validations wouldn't ever need to hit the database, even if they use promises, why we prefer a unit test. It should still work through all the motions, just stub out the part where it hits the database.

**Comment by xdarklight** [comment]
@mickhansen @sushantdhiman ah, now I see
I have updated this PR (and changed the title so it matches the patch contents)

File: test/unit/model/validation.test.js [hunk 1]
```diff
@@ -405,4 +405,163 @@ describe(Support.getTestDialectTeaser('InstanceValidator'), function() {
...
+      }
+    });
+
+    before(function () {
+      this.stub = sinon.stub(current, 'query', function () {
+        return new Promise(function (resolve) {
+          resolve(User.build({}));
+        });
+      });
+    });
```
Comment (mickhansen) [inline test/unit/model/validation.test.js]: Not a required change, just something you can consider for the future.

Stub could be simplified to:

`this.stub = sinon.stub(current, 'query').returns(Promise.resolve(User.build()));`
or
`this.sub = sinon.stub(current, 'query').resolves(User.build())` (assuming sinon as promised).

If model does not need to be a new instance each time.

File: test/unit/model/validation.test.js [hunk 2]
```diff
@@ -405,4 +405,163 @@ describe(Support.getTestDialectTeaser('InstanceValidator'), function() {
...
+      name: Sequelize.STRING
+    }, {
+      validate: {
+        customFn: function() {
+          var user = this;
+          return User.findAll().then(function() {
+            if (user.name === 'error') {
+              throw new Error('Error from model validation promise');
+            }
+          });
```
Comment (mickhansen) [inline test/unit/model/validation.test.js]: Could be simplified to:

```js
if (this.get('name') === 'error') {
  return Promise.reject(new Error('Error from model validation promise'));
}
return Promise.resolve();
```

**Comment by xdarklight** [comment]
@mickhansen thanks for reviewing this!
I just updated this PR - maybe you or @sushantdhiman could have another look at it

**Pull request merged**